osu.Game.Rulesets.Mania/Mods/IManiaRateAdjustmentMod.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) ppy Pty Ltd <contact@ppy.sh>. Licensed under the MIT Licence.
+// See the LICENCE file in the repository root for full licence text.
+
+using osu.Framework.Bindables;
+using osu.Game.Beatmaps;
+using osu.Game.Rulesets.Mania.Objects;
+using osu.Game.Rulesets.Mania.Scoring;
+using osu.Game.Rulesets.Mods;
+using osu.Game.Rulesets.Objects;
+using osu.Game.Rulesets.Scoring;
+
+namespace osu.Game.Rulesets.Mania.Mods
+{
+    /// <summary>
+    /// May be attached to rate-adjustment mods to adjust hit windows adjust relative to gameplay rate.
+    /// </summary>
+    /// <remarks>
+    /// Historically, in osu!mania, hit windows are expected to adjust relative to the gameplay rate such that the real-world hit window remains the same.
+    /// </remarks>
+    public interface IManiaRateAdjustmentMod : IApplicableToDifficulty, IApplicableToHitObject
+    {
+        BindableNumber<double> SpeedChange { get; }
+
+        HitWindows HitWindows { get; set; }
+
+        void IApplicableToDifficulty.ApplyToDifficulty(BeatmapDifficulty difficulty)
+        {
+            HitWindows = new ManiaHitWindows(SpeedChange.Value);
+            HitWindows.SetDifficulty(difficulty.OverallDifficulty);
+        }
+
+        void IApplicableToHitObject.ApplyToHitObject(HitObject hitObject)
+        {
+            switch (hitObject)
+            {
+                case Note:
+                    hitObject.HitWindows = HitWindows;
+                    break;
+
+                case HoldNote hold:
+                    hold.Head.HitWindows = HitWindows;
+                    hold.Tail.HitWindows = HitWindows;
+                    break;
```

Comment:
To avoid this, could we opt to have a nested structure where the multiplier is applied at the point of usage?

`hitObject.HitWindows = new AdjustedHitWindow(hitObject.HitWindows, SpeedChange.Value);`

It would sit better with me as if we add future alterations to `HitWindows` at hit objects it would play better with them.

Reply:
I'm extremely worried this is going to spiral out of control now...

Reply:
Can you elaborate what you're worried about specifically with the above suggestion? It seemed like a simple suggestion of a nice code quality improvement to me and not much more. Applying some composition is not generally something that would cause me to worry, so I'm guessing that you have a closer view on the issue and see potential issues with this.

Reply:
Just to clarify...

This immediately results in a ruleset API breaking change because I have to expose `HitWindows.GetRanges()` so that it can be referenced by a `RateAdjustedHitWindows`.

What happens when we have to eventually make this work for time ramping mods, but then we have _yet another_ source adjusting hit windows. Do we make this rate in `RateAdjustedHitWindows` a bindable? As it is, it's super easy to implement those mods - just replace the `HitWindows` property. What about when we (also eventually) want `BarHitErrorMeter` responding to real-time changes in `HitWindows`, how do we expose the "true hit windows" to the meter?

I wholly with this direction.

Reply:
I can understand this reasoning. Thanks for explaining.

I'm personally okay with a local implementation in this case, until it's proven we have to do something other than that.

Reply:
I think it's probably fine as-is.